### ISSUE TITLE
DllMain calls LoadLibrary for dxil.dll, could cause deadlock or crash

### ISSUE THREAD

**Issue by amaiorano** [microsoft/DirectXShaderCompiler#5985]
We've gotten a few Chrome crash reports where the GPU process spuriously hangs while loading `dxcompiler.dll` via LoadLibrary. The top of our call stacks where the hang is happening looks like:

```
(ntdll.dll + 0x0009d924)	NtCreateSection
(ntdll.dll + 0x0005fcfd)	LdrpMapDllNtFileName
(ntdll.dll + 0x0005fa7f)	LdrpMapDllFullPath
(ntdll.dll + 0x0005ed1e)	LdrpProcessWork
(ntdll.dll + 0x0001fb52)	LdrpLoadDllInternal
(ntdll.dll + 0x000173e3)	LdrpLoadDll
(ntdll.dll + 0x00016af3)	LdrLoadDll
```

This isn't much to go on, so I did some searching for what could cause DLL loading to hang, and from the [DllMain](https://learn.microsoft.com/en-us/windows/win32/dlls/dllmain) doc, I found [this DLL best practices](https://learn.microsoft.com/en-us/windows/win32/dlls/dynamic-link-library-best-practices) doc. In there, it mentions how when DllMain runs, it holds the "loader lock", and so must not perform any operation that does the same. In fact, at the top of the list of tasks you should never perform from within DllMain is:

* Call [LoadLibrary](https://learn.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-loadlibrarya) or [LoadLibraryEx](https://learn.microsoft.com/en-us/windows/desktop/api/LibLoaderAPI/nf-libloaderapi-loadlibraryexa) (either directly or indirectly). This can cause a deadlock or a crash.

Unfortunately, dxcompiler's DllMain does exactly that when it loads "dxil.dll", as it does [here](https://github.com/microsoft/DirectXShaderCompiler/blob/main/include/dxc/Support/dxcapi.use.h#L34C19-L34C19). Putting a breakpoint on that LoadLibrary call, we get this callstack:

```
dxcompiler.dll!dxc::DxcDllSupport::InitializeInternal(const char * dllName=0x00007ffe87dc1c87, const char * fnName=0x00007ffe87d01f80) Line 34	C++
dxcompiler.dll!dxc::DxcDllSupport::InitializeForDll(const char * dll=0x00007ffe87dc1c87, const char * entryPoint=0x00007ffe87d01f80) Line 95	C++
dxcompiler.dll!DxilLibInitialize() Line 29	C++
dxcompiler.dll!InitMaybeFail() Line 68	C++
dxcompiler.dll!DllMain(HINSTANCE__ * hinstDLL=0x00007ffe85340000, unsigned long Reason=1, void * reserved=0x0000000000000000) Line 104	C++
dxcompiler.dll!dllmain_dispatch(HINSTANCE__ * const instance=0x00007ffe85340000, const unsigned long reason=1, void * const reserved=0x0000000000000000) Line 281	C++
dxcompiler.dll!_DllMainCRTStartup(HINSTANCE__ * const instance=0x00007ffe85340000, const unsigned long reason=1, void * const reserved=0x0000000000000000) Line 335	C++
ntdll.dll!LdrpCallInitRoutine()	Unknown
ntdll.dll!LdrpInitializeNode()	Unknown
ntdll.dll!LdrpInitializeGraphRecurse()	Unknown
ntdll.dll!LdrpPrepareModuleForExecution()	Unknown
ntdll.dll!LdrpLoadDllInternal()	Unknown
ntdll.dll!LdrpLoadDll()	Unknown
ntdll.dll!LdrLoadDll()	Unknown
KernelBase.dll!LoadLibraryExW()	Unknown
```

Note that this callstack doesn't exactly match the one we're getting in our reports. Ours are hanging during `LdrpMapDllFullPath -> LdrpMapDllNtFileName -> NtCreateSection`, so we don't see DllMain here. However, it seems that DllMain is doing too much, including things it should not do, and I wonder if this is may still be the cause of our spurious hangs.

Any thoughts from Microsoft folks would be appreciated. Have you had reports or seen spurious hangs when dxcompiler.dll is loaded? I understand that most folks are probably using dxc.exe, so there's not much going on in there, but for applications that dynamically load dxcompiler.dll, I wonder if they've hit this before.

**Comment by llvm-beanz** [comment]
I marked this as tech-debt because we have a few loosely related but separate issues around the library initialization handling in decompiler.dll and dxil.dll. We may need to carve out some time to address some of these as part of replacing the dxcompiler interface.

**Comment by amaiorano** [comment]
As discussed offline, I think we should consider moving most (all?) of what DllMain is currently doing to `DxcCreateInstance*`. Not only would this avoid loader-lock issues, but it would also make loading the DLL faster, and should make using dxcompiler as a static library doable as DllMain won't need to be called at all (see [this issue](https://github.com/microsoft/DirectXShaderCompiler/issues/4766)).

Does anyone see a problem with this strategy? I was thinking of taking a stab at this myself.

**Comment by devshgraphicsprogramming** [comment]
@AnastaZluk read this and find out if it affects us

**Comment by amaiorano** [comment]
I took some time to investigate what would be required to move the loading of `dxil.dll` out of `dxcompiler.dll`'s `DllMain`. What I found is that [this commit](https://github.com/microsoft/DirectXShaderCompiler/commit/5a916c56d9244ce51512655f88ccbbc827f8b11d) is where it was moved to `DllMain`. Before that, dxil.dll would be loaded on-demand from `DxcCompiler::Compile`, and then unloaded after the function returned. The commit comment includes:

```
- Resolve lifetime of dxil.dll on dxcompiler.dll to guarantee access of memory generated from dxil.dll
```

Looking at the original code, I assume what this is referring to is the fact that when validation is enabled, the validator returns a "validated blob", and this is what's returned to the caller:

```cpp
              IFT(pValidator->Validate(
                pOutputBlob, DxcValidatorFlags_InPlaceEdit, &pValResult));
...
            CComPtr<IDxcBlob> pValidatedBlob;
            IFT(pValResult->GetResult(&pValidatedBlob));
            if (pValidatedBlob != nullptr) {
              std::swap(pOutputBlob, pValidatedBlob);
            }
            pValidator.Release();
```

So the memory for `pValidatedBlob` is allocated by `dxil.dll`, but when this function returns, `dxil.dll` may be unloaded (if it's the last reference to it), making this a dangling reference.

Thus, whatever change is made, dxil.dll must be loaded before compiling via `dxcompiler.dll`, and must remain loaded after the resulting blob is returned. One way to do this would be to tie the loading of dxil.dll to when `DxcCreateInstance` is called to create the compiler instance, and to unload `dxil.dll` when this compiler instance is destroyed, which I think could be done by doing the load/unload of `dxil.dll` in the ctor/dtor of the `DxcCompiler` class.

There are a few more things to look at, such as the fact that there are actually 3 DLLs that load `dxil.dll`: `dxcompiler`, `dxilconv`, and `dxrfallbackcompiller`. Also, I think this would not handle the case where the user uses `DxcCreateInstance` to create a validator object.

To summarize, I think this is a non-trivial change, and would require extensive testing. I'm not sure how well tested all the code paths related to the validator are presently.

Having said all that, I was thinking of a potential workaround for this loader lock problem. What if the client code makes sure to explicitly load `dxil.dll` before loading `dxcompiler.dll`? I think this should resolve any deadlock situation that may occur because `dxcompiler.dll`'s loading of `dxil.dll` would simply increase a ref count and not perform any actual loading. Does this make sense? If so, this is probably what we'll do in the short-term in Chrome.

**Comment by manvis** [comment]
Good thing I saw this, since it probably affects us as well. Also, if you're going to change how and where the `dxil.dll` is loaded in the future, please consider requirements of use cases such as ours as well.

I'm currently working on some shader hot-reload functionality for our software, which is a dynamically loaded plug-in for a different piece of software.

The base software does not ship with DXC/DXIL and we need to load it from a custom dependency directory that's next to the DLL that contains our plug-in. We can pass a full path to our copy of `dxcompiler.dll` when calling `LoadLibrary` inside our plug-in, but the `dxil.dll` is a bit of a problem for us.

Calling `SetDllDirectory` with our dependency directory before any `LoadLibrary` calls seems to do the right thing in cases that I've tried so far, but we'd feel a lot safer if we didn't need to mess with the DLL search paths, since that may affect other plug-ins. Since you're going to change how library loading is handled while fixing this, it would help us a lot if there was an explicit way to pass a loaded `dxil.dll` when creating an instance of the compiler. At very least, it would be nice if we could specify the full path to the desired `dxil.dll`, which DXC would then use to load it.

Thanks!

**Comment by amaiorano** [comment]
@manvis you may have seen the [commit above](https://github.com/chromium/chromium/commit/a737dac50a5ad043d75450482c2dca40791365e9) from Chrome that I landed that is workaround for this issue. We basically load dxil.dll ourselves before loading dxcompiler.dll. Since dxil.dll is already loaded, dxcompiler.dll's attempt to load it should basically be a no-op (except for adding a ref count in the loader). You may be able to do the same in your case, I think.

**Comment by manvis** [comment]
Yeah, I saw. Thanks! I'll move to the same approach since only adjusting the search path does nothing to prevent the possibility of the deadlock.

**Comment by damyanp** [comment]
@python3kgae - presumably this issue will go away once the code in dxil.dll is statically linked to dxcompiler.dll?
